baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/PreferCollectionConstructors.java

```diff
@@ -0,0 +1,348 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.baseline.errorprone;
+
+import com.google.auto.service.AutoService;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import com.google.errorprone.BugPattern;
+import com.google.errorprone.VisitorState;
+import com.google.errorprone.bugpatterns.BugChecker;
+import com.google.errorprone.fixes.SuggestedFix;
+import com.google.errorprone.fixes.SuggestedFixes;
+import com.google.errorprone.matchers.Description;
+import com.google.errorprone.matchers.Matcher;
+import com.google.errorprone.matchers.method.MethodMatchers;
+import com.sun.source.tree.ExpressionTree;
+import com.sun.source.tree.MethodInvocationTree;
+import com.sun.tools.javac.code.Types;
+import com.sun.tools.javac.tree.JCTree.JCExpression;
+import com.sun.tools.javac.tree.JCTree.JCMethodInvocation;
+import com.sun.tools.javac.util.List;
+import java.util.ArrayList;
+import java.util.EnumMap;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.IdentityHashMap;
+import java.util.LinkedHashMap;
+import java.util.LinkedHashSet;
+import java.util.LinkedList;
+import java.util.Map;
+import java.util.Set;
+import java.util.TreeMap;
+import java.util.TreeSet;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.CopyOnWriteArraySet;
+import java.util.stream.Collectors;
+
+@AutoService(BugChecker.class)
+@BugPattern(
+        name = "PreferCollectionConstructors",
+        link = "https://github.com/palantir/gradle-baseline#baseline-error-prone-checks",
+        linkType = BugPattern.LinkType.CUSTOM,
+        severity = BugPattern.SeverityLevel.SUGGESTION,
+        summary = "Since Java 7 the standard collection constructors should be used instead of the static factory "
+                + "methods provided by Guava.")
+public final class PreferCollectionConstructors extends BugChecker implements BugChecker.MethodInvocationTreeMatcher {
+
+    private static final long serialVersionUID = 1L;
+
+    private static final Matcher<ExpressionTree> NEW_ARRAY_LIST =
+            MethodMatchers.staticMethod()
+                    .onClass("com.google.common.collect.Lists")
+                    .named("newArrayList")
+                    .withParameters();
+
+    private static final Matcher<ExpressionTree> NEW_ARRAY_LIST_WITH_ITERABLE =
+            MethodMatchers.staticMethod()
+                    .onClass("com.google.common.collect.Lists")
+                    .named("newArrayList")
+                    .withParameters("java.lang.Iterable");
+
+    private static final Matcher<ExpressionTree> NEW_ARRAY_LIST_WITH_CAPACITY =
+            MethodMatchers.staticMethod()
+                    .onClass("com.google.common.collect.Lists")
+                    .namedAnyOf("newArrayListWithCapacity", "newArrayListWithExpectedSize")
```

Comment:
newArrayListWithExpectedSize docs suggest it will be deprecated soon, and recommend pre-sizing ourselves. I don't have a strong opinion either way on this one, especially considering `newArrayListWithExpectedSize` doesn't make guarantees about how much extra space it provides, I'd be okay with leaving it out of this PR so we can approach it separately with more careful rollout.
I'm a little worried that if we ban `newArrayListWithCapacity(int)` and not `newArrayListWithExpectedSize(int)` people may use the latter instead of `new ArrayList(int)`

Reply:
In a big repo I happened to have checked out there were 161 calls to this method. I looked through about 30 of them and 100% of those seemed like they should actually have been calls to `...WithCapacity` (or, rather `new ArrayList(int)`), i.e. none exceeded the given value.

I'd vote for pushing this through. If people prefer, I can look through the remaining calls.

Reply:
Personally I think I'd prefer to separate this one/leave it out so that we can ship an initial version of this errorprone rule which we know is an exact no-op, it seems a bit sneaky to have like 30 replacements which have exactly the same behaviour and one which gives you slightly different allocation/resizing behaviour.